src/Modules/Visualization/ShowField.cc

```diff
@@ -448,754 +538,293 @@ void GeometryBuilder::renderFacesLinear(
+
+      if (useNormals)
       {
-        /// Fix normal of Quads
-        if (points.size() == 4)
+        if (useFaceNormals)
         {
-          Vector edge1 = points[1] - points[0];
-          Vector edge2 = points[2] - points[1];
-          Vector edge3 = points[3] - points[2];
-          Vector edge4 = points[0] - points[3];
-
-          Vector norm = Cross(edge1, edge2) + Cross(edge2, edge3) + Cross(edge3, edge4) + Cross(edge4, edge1);
-
-          norm.normalize();
-
-          for (size_t i = 0; i < nodes.size(); i++)
-          {
-            normals[i] = invertNormals ? -norm : norm;
-          }
+          for(size_t i = 0; i < numNodes; ++i)
+            mesh->get_normal(normals[i], nodes[i]);
         }
-        /// Fix Normals of Tris
         else
         {
-          Vector edge1 = points[1] - points[0];
-          Vector edge2 = points[2] - points[1];
-          Vector norm = Cross(edge1, edge2);
-
-          norm.normalize();
-
-          for (size_t i = 0; i < nodes.size(); i++)
+          Vector norm;
+          if (useQuads)
           {
-            normals[i] = invertNormals ? -norm : norm;
+            Vector edge1 = points[1] - points[0];
+            Vector edge2 = points[2] - points[1];
+            Vector edge3 = points[3] - points[2];
+            Vector edge4 = points[0] - points[3];
+            norm = Cross(edge1, edge2) + Cross(edge2, edge3) + Cross(edge3, edge4) + Cross(edge4, edge1);
+            norm.normalize();
           }
-          //For future reference for a try at smoother rendering
-          /*
-          for (size_t i = 0; i < nodes.size(); i++)
+          else
           {
-          mesh->get_normal(normals[i], nodes[i]);
+            Vector edge1 = points[1] - points[0];
+            Vector edge2 = points[2] - points[1];
+            norm = Cross(edge1, edge2);
+            norm.normalize();
           }
-          */
-        }
-      }
-    }
 
-    // Default color single face no matter the element data.
-    if (colorScheme == ColorScheme::COLOR_UNIFORM)
-    {
-      addFaceGeom(points, normals, withNormals, iboIndex, iboBuffer, vboBuffer,
-        colorScheme, face_colors, state);
-    }
-
-    // Element data (Cells) so two sided faces.
-    else if (fld->basis_order() == 0 && mesh->dimensionality() == 3)
-    {
-      auto map = colorMap.get();
-      //two possible colors.
-      svals.resize(2);
-      vvals.resize(2);
-      tvals.resize(2);
-      face_colors = {{ColorRGB(1.,1.,1.),ColorRGB(1.,1.,1.)}};
+          for(size_t i = 0; i < numNodes; ++i)
+            normals[i] = norm;
+        }
 
-      VMesh::Elem::array_type cells;
-      mesh->get_elems(cells, *fiter);
+        if(invertNormals)
+          for(size_t i = 0; i < numNodes; ++i)
+            normals[i] = -normals[i];
+      }
 
-      if (fld->is_scalar())
+      if(useColorMap)
       {
-        fld->get_value(svals[0], cells[0]);
-
-        if (cells.size() > 1)
+        // Element data (Cells) so two sided faces.
```

Comment:
Part of this or a near-future refactoring should be to extract the coloring functions separately from the spatial functions